### PR TITLE
KAFKA-2570: commit offsets on rebalance/close when auto-commit is enabled

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

**Comment by hachikuji** [comment]
Build failure seems to be caused by KAFKA-2555, which (correctly) introduced the possibility of UnknownConsumerIdException being raised from commitSync(). I'll update the test to allow for this possibility.

**Comment by hachikuji** [comment]
This time failure looks like transient failure from KAFKA-2058.

File: core/src/test/scala/integration/kafka/api/ConsumerBounceTest.scala [hunk 1]
```diff
@@ -73,25 +76,38 @@ class ConsumerBounceTest extends IntegrationTestHarness with Logging {
...
+        assertEquals(consumer.position(tp), consumer.committed(tp).offset)
 
-      if (consumer.position(tp) == numRecords) {
-        consumer.seekToBeginning()
-        consumed = 0
+        if (consumer.position(tp) == numRecords) {
+          consumer.seekToBeginning()
+          consumed = 0
+        }
+      } catch {
```
Comment (guozhangwang) [inline core/src/test/scala/integration/kafka/api/ConsumerBounceTest.scala:108]: Add a TODO mentioning that this should be removed after we persist in ZK.

File: core/src/test/scala/integration/kafka/api/ConsumerBounceTest.scala [hunk 2]
```diff
@@ -73,25 +76,38 @@ class ConsumerBounceTest extends IntegrationTestHarness with Logging {
...
     this.producers.foreach(_.close)
 
-    var consumed = 0
+    var consumed = 0L
     val consumer = this.consumers(0)
-    consumer.subscribe(List(topic))
+
+    consumer.subscribe(List(topic), new ConsumerRebalanceListener {
+      override def onPartitionsAssigned(partitions: util.Collection[TopicPartition]) {
+        // it's possible the last commit before rebalancing can fail, so we must reset
```
Comment (guozhangwang) [inline core/src/test/scala/integration/kafka/api/ConsumerBounceTest.scala]: Ditto as below.

**Comment by guozhangwang** [comment]
LGTM overall, could you add a unit test for auto commit?

**Comment by hachikuji** [comment]
@guozhangwang Ready for another look when you have a chance.

**Comment by hachikuji** [comment]
The test failure looks to be unrelated. I've added a comment on KAFKA-2587 to see if that jira needs to be reopened.

**Pull request closed without merging**